### PR TITLE
mgr/cephadm: simplify tracking of daemon inventory

### PR DESCRIPTION
**Pull request by liewegas**
Avoid blocking/async behavior of _get_daemons() for (almost?) everything.

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
make check:

```pyhon
    def test_rgw_update_fail(self, _send_command, _get_connection, cephadm_module):
    
        with self._with_host(cephadm_module, 'host1'):
            with self._with_host(cephadm_module, 'host2'):
                ps = PlacementSpec(hosts=['host1'], count=1)
                c = cephadm_module.add_rgw(RGWSpec('realm', 'zone1', placement=ps))
                [out] = wait(cephadm_module, c)
                match_glob(out, "Deployed rgw.realm.zone1.host1.* on host 'host1'")
    
                ps = PlacementSpec(hosts=['host2'], count=1)
                c = cephadm_module.add_rgw(RGWSpec('realm', 'zone2', placement=ps))
                [out] = wait(cephadm_module, c)
                match_glob(out, "Deployed rgw.realm.zone2.host2.* on host 'host2'")
    
                with pytest.raises(OrchestratorError):
                    ps = PlacementSpec(hosts=['host1', 'host2'], count=2)
                    c = cephadm_module.update_rgw(RGWSpec('realm', 'zone1', placement=ps))
>                   [out] = wait(cephadm_module, c)
E                   ValueError: too many values to unpack (expected 1)
```

hm.

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by liewegas** [comment]
@sebastian-philipp I think this is ready (pending testing)

This does the scraping synchronously for the moment; that can be improved as a follow-on.

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1298,8 +1301,47 @@ def _refresh_host_daemons(self, host):
...
+            'last_update': di['last_update'],
+            'daemons': {},
+        }
+        for name, dd in di['daemons'].items():
+            j['daemons'][name] = dd.to_json()
+        self.set_store(DAEMON_CACHE_PREFIX + host, json.dumps(j))
+
+    def _daemon_cache_rm_host(self, host):
+        self.set_store(DAEMON_CACHE_PREFIX + host, None)
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]: Please move this to a `class DaemonCache`

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -764,33 +751,33 @@ def serve(self):
...
-            # we really need to do here is raise health alerts for individual
-            # hosts that fail and continue with the ones that do not fail.
-            if completion.exception is not None:
-                self.log.error('failed to refresh daemons: %s' % completion.exception)
+            cutoff = datetime.datetime.utcnow() - datetime.timedelta(
+                seconds=self.daemon_cache_timeout)
+            cutoffs = cutoff.strftime(DATEFMT)
+            self.log.debug('refreshing daemons, cutoff %s' % cutoffs)
+            failures = []
+            for host, di in self.daemon_cache.items():
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]: wondering, if we should move the cutoff calculation into the loop, as querying a few hundred hosts might take some time

**Comment by liewegas** [comment]
retest this please

**Pull request merged**